File: gui/backend/routes/findings.py
"""Cross-scan findings endpoints + per-finding actions."""
from __future__ import annotations

from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query

from gui.backend.schemas import FindingListResponse, GenericMessage
from gui.backend.services.report_reader import latest_report_for
from gui.backend.services.scan_manager import ScanManager, get_scan_manager


router = APIRouter(prefix="/api/findings", tags=["findings"])


def _all_findings(mgr: ScanManager) -> list[dict]:
    out: list[dict] = []
    for rec in mgr.list():
        report = latest_report_for(rec)
        if report is None:
            continue
        f = report.get("findings", {})
        if isinstance(f, dict):
            for tier in ("INTERESTING", "COMMON", "FALSE_ALARM"):
                for item in f.get(tier, []):
                    item = dict(item)
                    item.setdefault("scan_id", rec.id)
                    out.append(item)
        elif isinstance(f, list):
            for item in f:
                item = dict(item)
                item.setdefault("scan_id", rec.id)
                out.append(item)
    return out


@router.get("", response_model=FindingListResponse)
def list_findings(
    scan_id: Optional[str] = None,
    severity: Optional[str] = None,
    tier: Optional[str] = None,
    status: Optional[str] = None,
    limit: int = Query(default=200, ge=1, le=2000),
    offset: int = Query(default=0, ge=0),
    mgr: ScanManager = Depends(get_scan_manager),
) -> FindingListResponse:
    all_items = _all_findings(mgr)
    if scan_id:
        all_items = [f for f in all_items if f.get("scan_id") == scan_id]
    if severity:
        sev = severity.lower()
        all_items = [f for f in all_items
                     if (f.get("severity") or "").lower() == sev]
    if tier:
        t = tier.upper()
        all_items = [f for f in all_items
                     if (f.get("classification") or f.get("tier") or "").upper() == t]
    if status:
        all_items = [f for f in all_items if f.get("status") == status]
    total = len(all_items)
    return FindingListResponse(findings=all_items[offset:offset + limit],
                               total=total)


@router.get("/{finding_id}")
def get_finding(finding_id: str,
                mgr: ScanManager = Depends(get_scan_manager)) -> dict:
    for f in _all_findings(mgr):
        if f.get("id") == finding_id:
            return f
    raise HTTPException(status_code=404, detail="finding not found")


@router.post("/{finding_id}/exploit", response_model=GenericMessage)
def run_exploit(finding_id: str) -> GenericMessage:
    """Deliberately disabled. The frontend's ExploitRunner panel exists
    for future use; in production this endpoint must require explicit
    per-finding authorisation before doing anything."""
    raise HTTPException(
        status_code=501,
        detail=("exploit execution intentionally not implemented. "
                "Wiring this requires per-finding scope authorisation "
                "that HunterPy does not yet track."),
    )


@router.post("/{finding_id}/tags", response_model=GenericMessage)
def add_tags(finding_id: str) -> GenericMessage:
    # Tags are a UI-only concept until we add a SQLite table for them.
    return GenericMessage(message="tagging is client-side only in v2.1",
                          success=True)

File: gui/frontend/src/lib/api/findings.ts
import { apiClient } from './client'
import { ENDPOINTS } from './endpoints'
import { mock } from './mock'
import type { Finding, FindingStatus, Severity, Tier } from '@app-types/finding'

export interface FindingListParams {
  scan_id?: string
  severity?: Severity
  tier?: Tier
  status?: FindingStatus
  limit?: number
  offset?: number
}

export const findingsAPI = {
  async list(params?: FindingListParams) {
    if (apiClient.useMocks) {
      let all = await mock.findings()
      if (params?.scan_id)  all = all.filter((f) => f.scan_id === params.scan_id)
      if (params?.severity) all = all.filter((f) => f.severity === params.severity)
      if (params?.tier)
        all = all.filter((f) => (f.classification || f.tier) === params.tier)
      if (params?.status)   all = all.filter((f) => f.status === params.status)
      return { findings: all, total: all.length }
    }
    return apiClient.get<{ findings: Finding[]; total: number }>(
      ENDPOINTS.findings, { params }
    )
  },

  async listByScan(scanId: string, params?: Omit<FindingListParams, 'scan_id'>) {
    return findingsAPI.list({ ...params, scan_id: scanId })
  },

  async get(id: string): Promise<Finding> {
    if (apiClient.useMocks) {
      const all = await mock.findings()
      const found = all.find((f) => f.id === id)
      if (!found) throw new Error(`finding ${id} not found`)
      return found
    }
    return apiClient.get<Finding>(ENDPOINTS.finding(id))
  },

  async updateStatus(id: string, status: FindingStatus, notes?: string) {
    if (apiClient.useMocks) {
      const f = await findingsAPI.get(id)
      return { ...f, status, notes }
    }
    return apiClient.patch<Finding>(ENDPOINTS.finding(id), { status, notes })
  },

  async addTags(id: string, tags: string[]) {
    if (apiClient.useMocks) {
      const f = await findingsAPI.get(id)
      return { ...f, tags: [...new Set([...(f.tags || []), ...tags])] }
    }
    return apiClient.post<Finding>(ENDPOINTS.findingTags(id), { tags })
  },

  async runExploit(id: string) {
    if (apiClient.useMocks) {
      return {
        success: false,
        output: 'Mock mode: exploit execution disabled for safety.',
        timestamp: new Date().toISOString(),
      }
    }
    return apiClient.post<{ success: boolean; output: string; timestamp: string }>(
      ENDPOINTS.findingExploit(id)
    )
  },
}
